run/block.ts

```diff
@@ -1,174 +1,154 @@
+} | {
+  type: "interrupted";
+};
+
 export function createBlock<T>(
   frame: Frame,
   operation: () => Operation<T>,
 ): Block<T> {
-  let exhaustion: Exhausted<T> | undefined = void 0;
-  let observable = createObservable<IterationEvent<T>>();
+  let result: BlockResult<T> | void = void 0;
+  let results = createObservable<BlockResult<T>>();
+  let interruption = createObservable<void>();
+  let queue = createObservable<ReturnType<typeof $next>>();
   let controller = new AbortController();
-  let { notify } = observable;
+  let { signal } = controller;
+
+  signal.addEventListener("abort", () => interruption.notify());
 
   let enter = evaluate<() => void>(function* () {
     yield* shift<void>(function* (k) {
       return k;
     });
 
-    let iterator = lazy(() => operation()[Symbol.iterator]());
+    let thunks = queue.observe();
 
-    let exit = yield* shift<Exited<T>>(function* (k) {
-      let { signal } = controller;
-      signal.onabort = () => {
-        k({
-          type: "exited",
-          reason: "terminated",
-          result: { type: "rejected", error: new Error("halted") },
-        });
-      };
-
-      k({
-        type: "exited",
-        reason: "completed",
-        result: yield* reduce({
-          frame,
-          iterator,
-          notify,
-          signal,
-          start: $next(void 0),
-        }),
-      });
+    yield* reset(function* () {
+      result = yield* results.first();
     });
 
-    observable.notify(exit);
-
-    let exhausted: Exhausted<T> = {
-      type: "exhausted",
-      exit,
-      result: yield* (function* (): Computation<Result<void>> {
-        if (exit.reason === "completed") {
-          if (exit.result.type === "resolved") {
-            return { type: "resolved", value: void 0 };
-          } else {
-            return exit.result;
-          }
-        } else {
-          return yield* reduce<void>({
-            frame,
-            iterator: iterator as () => Iterator<Instruction, void>,
-            notify,
-            start: $abort(),
-          });
-        }
-      })(),
-    };
-
-    observable.notify(exhaustion = exhausted);
-  });
-
-  let block: Block<T> = create<Block<T>>("Block", {}, {
-    enter,
-    *abort() {
-      return yield* shift<Result<void>>(function* (k) {
-        yield* reset(function* () {
-          let exhausted = yield* block;
-          k(exhausted.result);
-        });
-        controller.abort();
-      });
-    },
-    observe: observable.observe,
+    yield* reset(function* () {
+      yield* interruption.first();
+      queue.notify($abort());
+    });
 
-    *[Symbol.iterator]() {
-      if (exhaustion) {
-        return exhaustion;
-      }
-      let observer = block.observe();
-      while (true) {
-        let event = yield* observer;
-        if (event.type === "exhausted") {
-          observer.drop();
-          return event;
-        }
-      }
-    },
-  });
-  return block;
-}
+    queue.notify($next(void 0));
```

Comment:
It's just another way of saying `undefined`. It's something I've been using stylistically lately so that the value aligns with the type.

```ts
let something: String | void = undefined;
```
vs
```ts
let something: String | void = void(0);
```

It somehow feels like the second initialization block is more true to the shape of the type, but it is 100% aesthetic.